### ISSUE TITLE
Add unnecessary_non_nullable_cast

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/linter#3146]
**Describe the rule you'd like to see implemented**

Related to https://www.reddit.com/r/dartlang/comments/s1if4e/dart_null_safety_vs_type_safety_dart_linter_in

There is currently no warning for when we are casting a value as non-nullable when that value could be nullable.
An example would be:

```dart
int? value = obj as int; // should probably be "as int?"
```

While this cast is technically correct, casting as non-nullable instead of nullable brings no value and is instead a source of errror

**Examples**

This can happen in various situations:

```dart
void function(int? value) {...}

function(obj as int); // warn

int? a = obj as int; // warn

int? fn() {
  return obj as int?; // warn
}
```

**Comment by srawlins** [comment]
No I don't think this case is covered by any static analysis.

**Comment by rrousselGit** [comment]
> Did you intend to include the ? on the next-to-last line?

No my bad! Fixed

> For example, is this fundamentally different than any other cast to a type that is narrower than the context would require, such as num n = obj as int;

That would make sense too yes!

**Comment by fishythefish** [comment]
Adding a data point: this lint would probably have prevented https://github.com/dart-lang/sdk/issues/52403 (here's the fix [CL](https://dart-review.googlesource.com/c/sdk/+/303420)).

I was about to file a lint request when I came across this issue. :)

Re: extending this lint to all casts - I was thinking about this when I was starting to write the lint request. My gut instinct is that nullability-related casts aren't inherently special, so extending to all casts should be fine. But I do suspect that a lot of the immediate value is in detecting bad migrations, so I would happily take a lint just restricted to this case if it turns out that the broader lint has some unforeseen edge cases.
